### PR TITLE
Add ParticleArray classes

### PR DESCRIPTION
**Pull request by atmyers**
These containers can store and process a collection of particles in the same way, whether the data is laid out as AoS or SoA. Also adds a test for this functionality.

This doesn't change anything in ParticleContainer yet - that will come later. 

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

File: Tests/Particles/ParticleArray/main.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    ParticleArrayType parray(1000);
+    auto particles = parray.get_particle_data();
+
+    amrex::ParallelFor( particles.size(), [=] AMREX_GPU_DEVICE (int i) noexcept
+    {
+        auto&& p = particles[i];
+	p.x() = 7.0;
+        p.y() = 8.0;
+	p.z() = 9.0;
```
Comment (WeiqunZhang) [inline Tests/Particles/ParticleArray/main.cpp]: Can you fix the indentation?

File: Src/Particle/AMReX_ParticleArray.H [hunk 2]
```diff
@@ -0,0 +1,272 @@
+#ifndef AMREX_PARTICLEARRAY_H_
+#define AMREX_PARTICLEARRAY_H_
+
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_ParticleArray.H:4]: Can we include `AMReX_Config.H` here?

File: Tests/Particles/ParticleArray/main.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    using ParticleArrayType = ParticleArray<amrex::Gpu::DeviceVector, ParticleType, DataLayoutTag>;
+
+    ParticleArrayType parray(1000);
+    auto particles = parray.get_particle_data();
+
+    amrex::ParallelFor( particles.size(), [=] AMREX_GPU_DEVICE (int i) noexcept
+    {
+        auto&& p = particles[i];
+	p.x() = 7.0;
+        p.y() = 8.0;
```
Comment (atmyers) [inline Tests/Particles/ParticleArray/main.cpp:41]:
```suggestion
       p.y() = 8.0;
```

File: Tests/Particles/ParticleArray/main.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    using ParticleType = Particle<double, double, double, int, int>;
+    using ParticleArrayType = ParticleArray<amrex::Gpu::DeviceVector, ParticleType, DataLayoutTag>;
+
+    ParticleArrayType parray(1000);
+    auto particles = parray.get_particle_data();
+
+    amrex::ParallelFor( particles.size(), [=] AMREX_GPU_DEVICE (int i) noexcept
+    {
+        auto&& p = particles[i];
+	p.x() = 7.0;
```
Comment (atmyers) [inline Tests/Particles/ParticleArray/main.cpp]:
```suggestion
	  p.x() = 7.0;
```

File: Tests/Particles/ParticleArray/main.cpp (hunk 1 above)
Comment (atmyers) [inline Tests/Particles/ParticleArray/main.cpp]:
```suggestion
	  p.z() = 9.0;
```

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**